File: src/lib/crdt/awset.test.ts
import { describe, expect, it } from "vitest";
import AWSet from "./awset";

const id1 = "a";
const id2 = "b";

describe("AWSet", () => {
    it("should be able to add an element", () => {
        const set = new AWSet();

        expect(set.value).toEqual(new Set());
        expect(set.add(id1, 1)).toEqual(new Set([1]));
    });

    it("should be able to add multiple elements", () => {
        const set = new AWSet();

        expect(set.value).toEqual(new Set());
        expect(set.add(id1, 1)).toEqual(new Set([1]));
        expect(set.add(id1, 2)).toEqual(new Set([1, 2]));
        expect(set.add(id1, 3)).toEqual(new Set([1, 2, 3]));
    });

    it("should be able to add an element multiple times", () => {
        const set = new AWSet();

        expect(set.value).toEqual(new Set());
        expect(set.add(id1, 1)).toEqual(new Set([1]));
        expect(set.add(id1, 1)).toEqual(new Set([1]));
        expect(set.add(id1, 1)).toEqual(new Set([1]));
    });

    it("should be able to remove an element", () => {
        const set = new AWSet();

        expect(set.value).toEqual(new Set());
        expect(set.add(id1, 1)).toEqual(new Set([1]));
        expect(set.remove(1)).toEqual(new Set());
    });

    it("should be able to remove multiple elements", () => {
        const set = new AWSet();

        expect(set.value).toEqual(new Set());
        expect(set.add(id1, 1)).toEqual(new Set([1]));
        expect(set.add(id1, 2)).toEqual(new Set([1, 2]));
        expect(set.add(id1, 3)).toEqual(new Set([1, 2, 3]));
        expect(set.remove(1)).toEqual(new Set([2, 3]));
        expect(set.remove(2)).toEqual(new Set([3]));
        expect(set.remove(3)).toEqual(new Set());
    });

    it("should be able to remove an element multiple times", () => {
        const set = new AWSet();

        expect(set.value).toEqual(new Set());
        expect(set.add(id1, 1)).toEqual(new Set([1]));
        expect(set.remove(1)).toEqual(new Set());
        expect(set.remove(1)).toEqual(new Set());
        expect(set.remove(1)).toEqual(new Set());
    });

    it("should be able to start with a value", () => {
        const set = new AWSet(
            new Set([
                [id1, 1, 1],
                [id1, 2, 2],
                [id1, 3, 3],
            ]),
        );

        expect(set.value).toEqual(new Set([1, 2, 3]));
    });

    it("should be able to merge with itself", () => {
        const set = new AWSet();

        set.add(id1, 1);
        set.add(id1, 2);

        expect(set.merge(set)).toEqual(new Set([1, 2]));
    });

    it("should be able to merge with another set", () => {
        const set1 = new AWSet();
        const set2 = new AWSet();

        set1.add(id1, 1);
        set1.add(id1, 2);
        set2.add(id2, 3);

        expect(set1.merge(set2)).toEqual(new Set([1, 2, 3]));
    });

    it("should be able to merge with another set with conflicting adds", () => {
        const set1 = new AWSet();
        const set2 = new AWSet();

        set1.add(id1, 1);
        set2.add(id2, 1);

        expect(set1.merge(set2)).toEqual(new Set([1]));
    });

    it("should be able to merge with another set with conflicting removes", () => {
        const set1 = new AWSet();
        const set2 = new AWSet();

        set1.add(id1, 1);
        set1.remove(1);
        set2.remove(1);

        expect(set1.merge(set2)).toEqual(new Set());
    });

    it("should be able to merge with another set with conflicting adds and removes", () => {
        const set1 = new AWSet();
        const set2 = new AWSet();

        set1.add(id1, 1);
        set1.remove(1);
        set2.add(id2, 1);

        expect(set1.merge(set2)).toEqual(new Set([1]));
    });

    it("should be able to be converted to JSON", () => {
        const set = new AWSet();

        set.add(id1, 1);
        set.add(id1, 2);
        set.add(id1, 3);

        const expected = {
            value: [
                [id1, 1, 1],
                [id1, 2, 2],
                [id1, 3, 3],
            ],
            dots: {
                [id1]: 3,
            },
        } as const;

        expect(set.toJSON()).toEqual(expected);
        expect(JSON.stringify(set)).toEqual(
            `{"value":[["${id1}",1,1],["${id1}",2,2],["${id1}",3,3]],"dots":{"${id1}":3}}`,
        );
    });
});

File: server/src/test/kotlin/pt/up/fe/sdle/crdt/AWSetTest.kt
package pt.up.fe.sdle.crdt

import kotlin.test.Test
import kotlin.test.assertEquals

private const val ID1 = "a"
private const val ID2 = "b"

internal class AWSetTest {
    @Test
    fun `AWSet - add element`() {
        val set = AWSet<Int>()

        assertEquals(setOf(), set.value)
        assertEquals(setOf(1), set.add(ID1, 1))
    }

    @Test
    fun `AWSet - add elements`() {
        val set = AWSet<Int>()

        assertEquals(setOf(), set.value)
        assertEquals(setOf(1), set.add(ID1, 1))
        assertEquals(
            setOf(
                1,
                2,
            ),
            set.add(ID1, 2),
        )
        assertEquals(
            setOf(1, 2, 3),
            set.add(ID1, 3),
        )
    }

    @Test
    fun `AWSet - add element multiple times`() {
        val set = AWSet<Int>()

        assertEquals(setOf(), set.value)
        assertEquals(setOf(1), set.add(ID1, 1))
        assertEquals(setOf(1), set.add(ID1, 1))
        assertEquals(setOf(1), set.add(ID1, 1))
    }

    @Test
    fun `AWSet - remove element`() {
        val set = AWSet<Int>()

        assertEquals(setOf(), set.value)
        assertEquals(setOf(1), set.add(ID1, 1))
        assertEquals(set.remove(1), setOf())
    }

    @Test
    fun `AWSet - remove elements`() {
        val set = AWSet<Int>()

        assertEquals(setOf(), set.value)
        assertEquals(setOf(1), set.add(ID1, 1))
        assertEquals(
            setOf(
                1,
                2,
            ),
            set.add(ID1, 2),
        )
        assertEquals(
            setOf(
                1,
                2,
                3,
            ),
            set.add(ID1, 3),
        )
        assertEquals(
            setOf(
                2,
                3,
            ),
            set.remove(1),
        )
        assertEquals(
            setOf(
                3,
            ),
            set.remove(2),
        )
        assertEquals(setOf(), set.remove(3))
    }

    @Test
    fun `AWSet - remove element multiple times`() {
        val set = AWSet<Int>()

        assertEquals(setOf(), set.value)
        assertEquals(setOf(1), set.add(ID1, 1))
        assertEquals(setOf(), set.remove(1))
        assertEquals(setOf(), set.remove(1))
        assertEquals(setOf(), set.remove(1))
    }

    @Test
    fun `AWSet - merge with itself`() {
        val set = AWSet<Int>()

        set.add(ID1, 1)
        set.add(ID1, 2)

        assertEquals(
            setOf(
                1,
                2,
            ),
            set.merge(set),
        )
    }

    @Test
    fun `AWSet - merge with another set`() {
        val set1 = AWSet<Int>()
        val set2 = AWSet<Int>()

        set1.add(ID1, 1)
        set1.add(ID1, 2)

        set2.add(ID2, 3)

        assertEquals(
            setOf(
                1,
                2,
                3,
            ),
            set1.merge(set2),
        )
    }

    @Test
    fun `should be able to merge with another set with conflicting adds`() {
        val set1 = AWSet<Int>()
        val set2 = AWSet<Int>()

        set1.add(ID1, 1)
        set1.add(ID1, 1)

        assertEquals(
            setOf(
                1,
            ),
            set1.merge(set2),
        )
    }

    @Test
    fun `should be able to merge with another set with conflicting removes`() {
        val set1 = AWSet<Int>()
        val set2 = AWSet<Int>()

        set1.add(ID1, 1)
        set1.remove(1)
        set2.remove(1)

        assertEquals(setOf(), set1.merge(set2))
    }

    @Test
    fun `should be able to merge with another set with conflicting adds and removes`() {
        val set1 = AWSet<Int>()
        val set2 = AWSet<Int>()

        set1.add(ID1, 1)
        set1.remove(1)
        set2.add(ID2, 1)

        assertEquals(setOf(1), set1.merge(set2))
    }
}
